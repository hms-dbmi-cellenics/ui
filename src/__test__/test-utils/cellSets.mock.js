const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
