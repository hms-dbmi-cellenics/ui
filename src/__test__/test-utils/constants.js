export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
