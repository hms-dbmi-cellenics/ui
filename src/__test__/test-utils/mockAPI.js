import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
