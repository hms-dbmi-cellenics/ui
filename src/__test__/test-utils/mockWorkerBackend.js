import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
