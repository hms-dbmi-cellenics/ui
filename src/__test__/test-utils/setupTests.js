import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
