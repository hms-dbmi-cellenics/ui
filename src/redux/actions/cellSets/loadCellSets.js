import fetchAPI from 'utils/http/fetchAPI';
import handleError from 'utils/http/handleError';
import {
  CELL_SETS_LOADED, CELL_SETS_LOADING, CELL_SETS_ERROR,
} from 'redux/actionTypes/cellSets';
import endUserMessages from 'utils/endUserMessages';

const loadCellSets = (experimentId, forceReload = false) => async (dispatch, getState) => {
  const {
    loading, error, updatingClustering, initialLoadPending,
  } = getState().cellSets;

  const loadingBlocked = loading || updatingClustering;
  const requiresLoading = initialLoadPending || error;

  const shouldLoad = requiresLoading && !loadingBlocked;

  if (!shouldLoad && !forceReload) {
    return;
  }

  dispatch({
    type: CELL_SETS_LOADING,
  });

  try {
    const data = await fetchAPI(`/v2/experiments/${experimentId}/cellSets`);

    dispatch({
      type: CELL_SETS_LOADED,
      payload: {
        experimentId,
        data: data.cellSets,
      },
    });
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_CELL_SETS);

    dispatch({
      type: CELL_SETS_ERROR,
      payload: { error: errorMessage },
    });
  }
};

export default loadCellSets;
