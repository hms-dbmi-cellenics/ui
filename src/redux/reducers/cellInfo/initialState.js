const initialState = {
  component: null,
  cellId: null,
  geneName: null,
  trackCluster: null,
  focus: {
    store: 'cellSets',
    key: 'louvain',
  },
  groupedTrack: 'louvain',
  selectedTracks: ['louvain'],
};

export default initialState;
