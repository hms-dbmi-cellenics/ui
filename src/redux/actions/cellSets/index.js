import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import runCellCycleScoring from './runCellCycleScoring';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetSelected,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
  runCellCycleScoring,
};
