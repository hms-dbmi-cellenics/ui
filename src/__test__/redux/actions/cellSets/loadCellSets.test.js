import _ from 'lodash';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import loadCellSets from 'redux/actions/cellSets/loadCellSets';
import initialState from 'redux/reducers/cellSets/initialState';

import '__test__/test-utils/setupTests';
import { CELL_SETS_ERROR, CELL_SETS_LOADED, CELL_SETS_LOADING } from 'redux/actionTypes/cellSets';

enableFetchMocks();
const mockStore = configureStore([thunk]);

describe('loadCellSets action', () => {
  const experimentId = '1234';

  const cellSets = {
    name: 'one', color: '#ff0000', cellIds: new Set([1, 2, 3]),
  };

  const experimentSettings = {
    info: {
      sampleIds: [],
    },
  };

  beforeEach(() => {
    const response = new Response(JSON.stringify({ cellSets }));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  it('Does not dispatch on normal operation', async () => {
    const store = mockStore({
      cellSets: {
        initialLoadPending: false, loading: false, error: false, updatingClustering: false,
      },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches on force reload', async () => {
    const store = mockStore({ cellSets: { loading: false, error: false }, experimentSettings });
    await store.dispatch(loadCellSets(experimentId, true));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
  });

  it('Dispatches on initial load pending false and error state', async () => {
    const store = mockStore({
      cellSets: { initialLoadPending: false, loading: false, error: true },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches a loaded action when run with the initial state.', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches an error condition if fetch fails', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });

    fetchMock.resetMocks();
    fetchMock.mockReject(new Error('some weird error that happened'));

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_ERROR]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Uses V2 URL when using API version V2', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    const fetchUrl = fetchMock.mock.calls[0][0];

    expect(fetchUrl).toEqual('http://localhost:3000/v2/experiments/1234/cellSets');
  });
});
