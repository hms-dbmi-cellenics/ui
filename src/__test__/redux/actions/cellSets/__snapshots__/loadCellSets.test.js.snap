// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCellSets action Dispatches a loaded action when run with the initial state. 1`] = `
Array [
  Object {
    "type": "cellSets/loading",
  },
  Object {
    "payload": Object {
      "data": Object {
        "cellIds": Object {},
        "color": "#ff0000",
        "name": "one",
      },
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;

exports[`loadCellSets action Dispatches an error condition if fetch fails 1`] = `
Array [
  Object {
    "type": "cellSets/loading",
  },
  Object {
    "payload": Object {
      "error": "We couldn't get the list of cell sets.",
    },
    "type": "cellSets/error",
  },
]
`;

exports[`loadCellSets action Dispatches on initial load pending false and error state 1`] = `
Array [
  Object {
    "type": "cellSets/loading",
  },
  Object {
    "payload": Object {
      "data": Object {
        "cellIds": Object {},
        "color": "#ff0000",
        "name": "one",
      },
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;
